src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+
+int ngx_http_lua_ffi_set_semaphore_threshold(ngx_int_t threshold);
+static ngx_http_lua_semaphore_t *ngx_http_lua_alloc_sem(ngx_log_t *log);
+void ngx_http_lua_cleanup_sem(void *data);
+static void ngx_http_lua_free_sem(ngx_http_lua_semaphore_t *sem, ngx_log_t *log);
+static ngx_int_t ngx_http_lua_semaphore_resume(ngx_http_request_t *r);
+int ngx_http_lua_ffi_sem_new(ngx_http_lua_semaphore_t **psem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_post(ngx_http_lua_semaphore_t *sem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_wait(ngx_http_request_t *r, ngx_http_lua_semaphore_t *sem,
+                              int time, u_char *errstr, size_t *errlen);
+static void ngx_http_lua_ffi_semaphore_cleanup(void *data);
+void ngx_http_lua_ffi_semaphore_handler(ngx_event_t *ev);
+static void ngx_http_lua_sem_timeout_handler(ngx_event_t *ev);
+void ngx_http_lua_ffi_sem_gc(ngx_http_lua_semaphore_t *sem);
+
+static ngx_http_lua_semaphore_t *ngx_http_lua_sem_free_list = NULL;
+static ngx_int_t ngx_http_lua_semaphore_threshold = 0;
+static ngx_int_t ngx_http_lua_sem_free_count = 0;
+
+
+int
+ngx_http_lua_ffi_set_semaphore_threshold(ngx_int_t threshold)
+{
+    ngx_http_lua_semaphore_threshold = threshold;
+    return (int)ngx_http_lua_semaphore_threshold;
+}
+
+
+int
+ngx_http_lua_ffi_get_semaphore_threshold()
+{
+    return (int)ngx_http_lua_semaphore_threshold;
+}
+
+
+static ngx_http_lua_semaphore_t*
+ngx_http_lua_alloc_sem(ngx_log_t *log)
+{
+    ngx_http_lua_semaphore_t *sem;
+
+    if (ngx_http_lua_sem_free_list) {
+         sem = ngx_http_lua_sem_free_list;
+         ngx_http_lua_sem_free_list = ngx_http_lua_sem_free_list->next;
+         ngx_http_lua_sem_free_count--;
+
+    } else {
+
+        sem = ngx_alloc(sizeof(ngx_http_lua_semaphore_t), log);
+        if (sem == NULL) {
+            return NULL;
+        }
+    }
+
+    dd("ngx_http_lua_alloc_sem sem:%p", sem);
+
+    return sem;
+}
+
+
+void
+ngx_http_lua_cleanup_sem(void *data)
+{
+    dd("ngx_http_lua_cleanup_sem");
+
+    ngx_http_lua_semaphore_t * sem;
+
+    while ((sem=ngx_http_lua_sem_free_list) != NULL)
+    {
+         ngx_http_lua_sem_free_list = ngx_http_lua_sem_free_list->next;
+         ngx_free(sem);
+    }
+}
+
+
+static void
+ngx_http_lua_free_sem(ngx_http_lua_semaphore_t *sem, ngx_log_t *log)
+{
+    ngx_int_t n;
+
+    dd("ngx_http_lua_free_sem sem:%p", sem);
+
+    sem->next = ngx_http_lua_sem_free_list;
+    ngx_http_lua_sem_free_list = sem;
+    ngx_http_lua_sem_free_count++;
+
+    if (ngx_http_lua_sem_free_count >= ngx_http_lua_semaphore_threshold) {
+
+         n = ngx_http_lua_sem_free_count >> 1;
+         ngx_http_lua_sem_free_count -= n;
+
+         while (n--) {
+             sem = ngx_http_lua_sem_free_list;
+             ngx_http_lua_sem_free_list = ngx_http_lua_sem_free_list->next;
+             ngx_free(sem);
+         }
+
+         dd("ngx_http_lua_free_sem sem_count:%lld",
+                 (long long int) ngx_http_lua_sem_free_count);
```

Comment:
Style: continued argument list should be aligned right after the `(` char in the previous line.